README
#! /usr/bin/less
################################################################################

Program : log-malloc2 (library)
Version	: 0.3.0
Purpose	: Malloc logging library with backtrace and byte-exact memory
		tracking.
License	: GNU GPL v3 (see file COPYING)
	  GNU LGPL v3 (see file COPYING-LGPL)
Author	: Samuel Behan <_samuel_._behan_(at)_dob_._sk> (c) 2011-2014
Web	: http://devel.dob.sk/log-malloc2
	  http://blog.dob.sk/category/devel/log-malloc2 (howto, tutorial)

################################################################################

---------
- ABOUT -
---------

     log-malloc2 is library for tracking memory functions calls, that can be
     preloaded using LD_PRELOAD or even directly linked to your program.

     It also provides simple api to programaticaly check current memory usage
     (bytes requested) intended for testing purposes.

     ---
     This library is partialy based on log-malloc library, hence the name
     log-malloc2 - and I'm too lazy to figure out better name. I've taken some
     ideas from this library, but it's completely rewriten to make backtracing
     work also on 64bit system, and to provide also additional usefull informations
     regarding allocated memory.


------------
- FEATURES -
------------

     * logging to file descriptor 1022 (if opened)
     * call stack backtrace (using GNU backtrace() if available)
     * requested memory tracking (byte-exact)
     * allocated memory tracking (byte-exact - using malloc_usable_size())
     * process memory status tracking (from /proc/self/statm)
     * call counting
     * thread safe (used also to prevent log mixing)
     * optional C API for runtime memory usage checking


---------
- USAGE -
---------

        log-malloc -o /tmp/program.log command args

           OR

	LD_PRELOAD=./liblog-malloc2.so command args ... 1022>/tmp/program.log

     This generates log-malloc trace file, tracking all memory operations and
     associated stack backtraces that helps to identify source file/line that
     made that memory operation.

     Having program compiled with debug information and -rdynamic makes backtrace
     call stack better readable. Aferwards, you can use backtrace2line script or
     addr2line tool from GNU binutils to convert backtrace addresses into exact
     source code line numbers (if using addr2line you have to deal with ASLR
     yourselves).

     If trace file descriptor is not opened for writing on library initialisation,
     trace output is automatically disabled, and should not further harm program
     performance anyhow.

     This package provides also additional scripts, that will help you to deal
     with log-malloc2 trace:
 
     	* backtrace2line
		Script to automatically convert backtrace in files and line numbers.
		Can also deal with ASLR randomized addresses.

	* log-malloc-findleak
		Script to discover possible program memory leaks from trace file

	* log-malloc-trackusage
		Script to track program memory usage over time.

     This scripts can be also used as perl packages, because they export functions
     to parse and analyse trace file or convert backtraces.


---------
- C API -
---------

     log-malloc2 provides also simple C API functions (ideal for testing) that can be used
     if it is directly linked to a program at compile/link time.
     Warning: Such a usage is not recommended in production environment! Memory tracking
         and call counting use memory bariers, that may degrade application performance.

     size_t log_malloc_get_usage(void)

	Get actual program memory usage in bytes

     void log_malloc_trace_enable(void)

	Enable trace messages to be printed to trace fd.

     void log_malloc_trace_disable(void)

	Disable trace messages.

     int log_malloc_trace_printf(const char *fmt, ...)

	Printf smth. to trace fd (message size is limited to 1024 bytes).

     LOG_MALLOC_SAVE(name, trace) [MACRO]

        Creates savepoint with given _name_ that stores actual memory usage.
        If _trace_ is true, message will be logged to trace fd.

     LOG_MALLOC_UPDATE(name, trace) [MACRO]

        Updates actual memory usage in savepoint with given _name_.
        If _trace_ is true, trace message will be logged to trace fd.

     LOG_MALLOC_COMPARE(name, trace) [MACRO]

        Compares actual memory usage with the saved one under given _name_.
        If _trace_ is true, trace message will be logged to trace fd.
        Call returns memory usage difference (size_t).

     LOG_MALLOC_ASSERT(name, iter) [MACRO]

	ASSERT with fail, if actual memory usage differs from the one saved in savepoint.
	_iter_ can specify that assertion should be checked first after
	given number of LOG_MALLOC_SAVE() iterations.

     LOG_MALLOC_NDEBUG [MACRO]
        If defined, above macros will generate no code.

--------------------
- PERFORMANCE TIPS -
--------------------

    * Log to tmpfs, or other FS that handles write operation effectively
        If traced application intensively allocates memory, consider logging to tmpf
	because writting to trace fd is the slowest operation, that is in addition
	protected by a mutex, thus serializing multithreaded memory allocations.


---------------------------------
- TRACKED FUNCTIONS & INTERNALS -
---------------------------------

	* malloc
	* realloc
	* calloc
		Note: library changes passed parameters from calloc(nmemb, size)
		  to calloc(1, nmemb*size). Allocated memory is exactly the same
		  but there might(?) appear some problem due to this change.
	* memalign
	* posix_memalign
	* valloc
	* free
		Note: call to free are not backtraced, also free(NULL) calls are
		  completely ignored.

     Library allocates additional 32 bytes (due to structure alignment) per every allocated
     memory. This is necessary to make byte-exact memory tracking possible.
     This memory is placed in front of allocated memory and stores size of allocated
     memory (1 x size_t) and check bits (1 x size_t) for allocated size number and
     real-aligned allocated memory size (1 x size_t).
     These data are also used to distinguish memory allocated by this library,
     and memory that has been allocated by something else (whatever it was, you can
     identify it by appended '!f' when free()-ing).

     This functionality makes log-malloc2 library less resistant against random
     memory overwrites, that will destroy library internal data (especially by
     realloc/free), but incorrect memory overwrites are anyway bad, and this
     makes them only better recognizable.

     Library uses malloc_usable_size() function (if available) to determine how much memory
     has been really allocated by libc. Allocated memory differs from requested memory,
     because alocated memory blocks are usualy aligend to some power of 2.

     Note: Allocated memory status (MEM-STATUS) might and will differ from
       numbers you will see in top/ps or statm data, because memory is allocated
       in blocks that are usually greater than requested memory, but it might
       be relatively close to MEM-STATUS-ALIGNED.
       DuckDuckGo it ;)


----------
- OUTPUT -
----------

     Logfile has following structure

	+ FUNCTION MEM-CHANGE MEM-IN? MEM-OUT? (FUNCTION-PARAMS) [MEM-STATUS:MEM-STATUS-USABLE] #STATM-DATA
	BACKTRACE
	...
	...
	+ FUNCTION....
	...
	# ADDITIONAL-DATA
	...

     Legend:

	FUNCTION		- called function name (malloc/realloc/free...)
	MEM-CHANGE		- bytes allocated/deallocated
	MEM-IN			- (optional) input memory pointer
	MEM-OUT			- (optional) (re)allocated memory pointer
	MEM-STATUS		- actually allocated memory using functions
	MEM-STATUS-USABLE	- sum of allocated memory reported by malloc_usable_size
	FUNCTION-PARAMS		- parameter that has been passed to function
	STATM-DATA		- copy of /proc/self/statm content
	ADDITIONAL-DATA		- additional runtime data, like PID, CWD, MAPS content


------------------
- EXAMPLE OUTPUT -
------------------

 $ log-malloc.pl -o - ./examples/leak-01

 *** log-malloc trace-fd = 1022 ***
 
 + malloc 53 0x7f9bff564080 [85:160]!
 + calloc 1182 0x7f9bff5640e0 [1267:1384] (1182 1)!
 + malloc 53 0x7f9bff5645b0 [1320:1472]!
 + malloc 56 0x7f9bff564610 [1376:1560]!
 + calloc 360 0x7f9bff564670 [1736:1952] (15 24)!
 + calloc 32 0x7f9bff564030 [32:72] (1 32) #3183 168 131 1 0 84 0
 /lib64/libdl.so.2(+0x1960)[0x7f9bfcfa8960]
 /lib64/libdl.so.2(dlsym+0x5a)[0x7f9bfcfa843a]
 /home/sam/devel/log-malloc2/scripts/../.libs/liblog-malloc2.so(+0x12e3)[0x7f9bfd7882e3]
 /lib64/ld-linux-x86-64.so.2(+0xfa0b)[0x7f9bfd99ba0b]
 /lib64/ld-linux-x86-64.so.2(+0xfb1c)[0x7f9bfd99bb1c]
 /lib64/ld-linux-x86-64.so.2(+0x140a)[0x7f9bfd98d40a]
 /lib64/libdl.so.2(+0x13a0)[0x7f9bfcfa83a0]
 # PID 23451
 # EXE /home/sam/devel/log-malloc2/examples/leak-01
 # CWD /home/sam/devel/log-malloc2
 + INIT [1736:1952] malloc=3 calloc=3 realloc=0 memalign=0/0 valloc=0 free=0
 + malloc 2000 0x7f9bff564800 [3736:3992] #3183 168 131 1 0 84 0
 ./examples/leak-01(main+0x32)[0x7f9bfdbb1af2]
 /lib64/libc.so.6(__libc_start_main+0x11b)[0x7f9bfd3f01cb]
 ./examples/leak-01(+0x9a9)[0x7f9bfdbb19a9]
 [0x0]
 + malloc 100 0x7f9bff565000 [3836:4128] #3183 168 131 1 0 84 0
 ./examples/leak-01(main+0x40)[0x7f9bfdbb1b00]
 /lib64/libc.so.6(__libc_start_main+0x11b)[0x7f9bfd3f01cb]
 ./examples/leak-01(+0x9a9)[0x7f9bfdbb19a9]
 [0x0]
 + free -100 0x7f9bff565000 [3736:3992] #3183 168 131 1 0 84 0
 + FINI [3736:3992] malloc=5 calloc=3 realloc=0 memalign=0/0 valloc=0 free=1
 # FILE /proc/self/maps
 7f9bfcd90000-7f9bfcda6000 r-xp 00000000 fe:02 911050                     /usr/lib64/gcc/x86_64-pc-linux-gnu/4.8.3/libgcc_s.so.1
 7f9bfcda6000-7f9bfcfa5000 ---p 00016000 fe:02 911050                     /usr/lib64/gcc/x86_64-pc-linux-gnu/4.8.3/libgcc_s.so.1
 7f9bfcfa5000-7f9bfcfa6000 r--p 00015000 fe:02 911050                     /usr/lib64/gcc/x86_64-pc-linux-gnu/4.8.3/libgcc_s.so.1
 7f9bfcfa6000-7f9bfcfa7000 rw-p 00016000 fe:02 911050                     /usr/lib64/gcc/x86_64-pc-linux-gnu/4.8.3/libgcc_s.so.1
 7f9bfcfa7000-7f9bfcfaa000 r-xp 00000000 fe:02 819448                     /lib64/libdl-2.19.so
 7f9bfcfaa000-7f9bfd1a9000 ---p 00003000 fe:02 819448                     /lib64/libdl-2.19.so
 7f9bfd1a9000-7f9bfd1aa000 r--p 00002000 fe:02 819448                     /lib64/libdl-2.19.so
 7f9bfd1aa000-7f9bfd1ab000 rw-p 00003000 fe:02 819448                     /lib64/libdl-2.19.so
 7f9bfd1ab000-7f9bfd1c5000 r-xp 00000000 fe:02 818913                     /lib64/libpthread-2.19.so
 7f9bfd1c5000-7f9bfd3c5000 ---p 0001a000 fe:02 818913                     /lib64/libpthread-2.19.so
 7f9bfd3c5000-7f9bfd3c6000 r--p 0001a000 fe:02 818913                     /lib64/libpthread-2.19.so
 7f9bfd3c6000-7f9bfd3c7000 rw-p 0001b000 fe:02 818913                     /lib64/libpthread-2.19.so
 7f9bfd3c7000-7f9bfd3cb000 rw-p 00000000 00:00 0
 7f9bfd3cb000-7f9bfd57d000 r-xp 00000000 fe:02 819451                     /lib64/libc-2.19.so
 7f9bfd57d000-7f9bfd77d000 ---p 001b2000 fe:02 819451                     /lib64/libc-2.19.so
 7f9bfd77d000-7f9bfd781000 r--p 001b2000 fe:02 819451                     /lib64/libc-2.19.so
 7f9bfd781000-7f9bfd783000 rw-p 001b6000 fe:02 819451                     /lib64/libc-2.19.so
 7f9bfd783000-7f9bfd787000 rw-p 00000000 00:00 0
 7f9bfd787000-7f9bfd78b000 r-xp 00000000 fe:02 14619                      /home/sam/devel/log-malloc2/.libs/liblog-malloc2.so.1.0.0
 7f9bfd78b000-7f9bfd98a000 ---p 00004000 fe:02 14619                      /home/sam/devel/log-malloc2/.libs/liblog-malloc2.so.1.0.0
 7f9bfd98a000-7f9bfd98b000 r--p 00003000 fe:02 14619                      /home/sam/devel/log-malloc2/.libs/liblog-malloc2.so.1.0.0
 7f9bfd98b000-7f9bfd98c000 rw-p 00004000 fe:02 14619                      /home/sam/devel/log-malloc2/.libs/liblog-malloc2.so.1.0.0
 7f9bfd98c000-7f9bfd9ae000 r-xp 00000000 fe:02 818908                     /lib64/ld-2.19.so
 7f9bfdb99000-7f9bfdb9d000 rw-p 00000000 00:00 0
 7f9bfdbad000-7f9bfdbae000 rw-p 00000000 00:00 0
 7f9bfdbae000-7f9bfdbaf000 r--p 00022000 fe:02 818908                     /lib64/ld-2.19.so
 7f9bfdbaf000-7f9bfdbb0000 rw-p 00023000 fe:02 818908                     /lib64/ld-2.19.so
 7f9bfdbb0000-7f9bfdbb1000 rw-p 00000000 00:00 0
 7f9bfdbb1000-7f9bfdbb2000 r-xp 00000000 fe:02 15340                      /home/sam/devel/log-malloc2/examples/leak-01
 7f9bfddb1000-7f9bfddb2000 r--p 00000000 fe:02 15340                      /home/sam/devel/log-malloc2/examples/leak-01
 7f9bfddb2000-7f9bfddb3000 rw-p 00001000 fe:02 15340                      /home/sam/devel/log-malloc2/examples/leak-01
 7f9bff564000-7f9bff585000 rw-p 00000000 00:00 0                          [heap]
 7fffb5471000-7fffb5493000 rw-p 00000000 00:00 0                          [stack]
 7fffb54cf000-7fffb54d1000 r-xp 00000000 00:00 0                          [vdso]
 ffffffffff600000-ffffffffff601000 r-xp 00000000 00:00 0                  [vsyscall]


NEWS
0.3.0	2014-11-01

IMPROVEMENTS

	* see Changelog

0.2.3	2013-09-24

BUGFIX

	* fixed pointer formating
	* automake tools update

0.2.3	2012-03-29

IMPROVEMENTS

	* improved library compatibility
	* detect if backtrace() is available
	* some docu fixes

0.2.2	2012-03-26

IMPROVEMENTS

	* track really allocated memory using malloc_usable_size()

0.2.1	2012-02-14

BUGFIX + IMPROVEMENTS

	* tracking more functions
	* fixed possible segfault if alloc fails

0.2.0	2012-02-08

INITAL RELEASE

	* initial public release
	* autoconf/automake packaged
	* minimal documentation in README file




scripts/backtrace2line.pl
#!/usr/bin/perl -w
#
# log-malloc2 / backtrace2line
#	Translate backtrace() output into functions, file names and line numbers (supports ASLR)
#
# Author: Samuel Behan <_samuel_._behan_(at)_dob_._sk> (C) 2013-2014
#
# License: GNU GPLv3 (http://www.gnu.org/licenses/gpl.html)
#
# Web:
#	http://devel.dob.sk/log-malloc2
#	http://blog.dob.sk/category/devel/log-malloc2 (howto, tutorials)
#	https://github.com/samsk/log-malloc2 (git repo)
#
#
package log_malloc::backtrace2line;

use strict;
use Cwd;
use Getopt::Long;
use Pod::Usage;
use Data::Dumper;
use File::Basename;

# VERSION
our $VERSION = '0.3';

# CONFIGS
our $VERBOSE = 0;
my @ADDR2LINE = ("addr2line", "-C");
my $LINUX_ALSR_CONFIG = "/proc/sys/kernel/randomize_va_space";

# EXEC
sub main(@);
exit(main(@ARGV)) if(!caller());

#
# INTERNAL FUNCTIONS
#

sub _readFile($)
{
	my ($fd) = @_;

	my @arr;
	while(my $line = <$fd>)
	{
		chomp($line);

		# read till line with only '.' if reading from FILE handle
		last
			if($line eq ".");

		push(@arr, $line);
	}
	return @arr;
}

sub verbose($$)
{
	my ($lev, $msg) = @_;

	return print(STDERR $msg)
		if($lev <= $VERBOSE);
	return 0;
}

sub read_mapsFile($$)
{
	my ($mapsFile, $pid) = @_;

	# mapsFile can be an ARRAY (for moduline call)
	if(ref($mapsFile) eq 'ARRAY')
	{
		return @$mapsFile;
	}
	elsif($pid)
	{
		$mapsFile = sprintf("/proc/%d/maps", $pid);
	}
	elsif(ref($mapsFile))
	{
		die("[programming error] invalid mapsFile=$mapsFile argument !");
	}

	# read from user supplied maps file
	my $fd;
	return ()
		if(!$mapsFile || (!open($fd, $mapsFile)));

	my @maps = _readFile($fd);
	close($fd);

	return @maps;
}

sub get_libmap(@)
{
	my (@maps) = @_;

	no warnings 'portable';
	my %libs;
	foreach my $line (@maps)
	{
		my ($range, $perms, $offset, $dev, $inode, $path) = split(/\s+/o, $line);
		my ($addr_from, $addr_to) = split(/-/o, $range);

		# need only executable sections
		next
			if($perms !~ /x/o || !$path || $path =~ /^\[\w+\]$/o);

		# unhex
		$addr_from = hex($addr_from);

		my $libname = basename($path);
		$libs{$libname} = $addr_from;
		$libs{'-' . $libname} = $path;
	}
	return %libs;
}

sub get_wd($$)
{
	my ($wd, $pid) = @_;

	return $wd || undef
		if($wd || !$pid);

	my $lnk = sprintf("/proc/%d/cwd", $pid);
	return readlink($lnk);
}

#
# PUBLIC FUNCTIONS
#

# addr2lime(@executable, @symbols): @resolved_symbols
sub addr2line($@)
{
	my ($exe, @symbols) = @_;

	my @cmd = (@ADDR2LINE, "-f", "-e", $exe, @symbols);

	# XXX: improve this
	my @lines = `@cmd`;
	warn("\t - addr2line filed while executing '@cmd'\n"), return ()
		if($? != 0);

	my %addr;
	my $symIdx = 0;
	chomp(@lines);
	for(my $ii = 0; $ii <= $#lines; $ii++)
	{
		my $line = $lines[$ii];
		my $sym = $symbols[$symIdx];

		# every 1st line is function
		my $function = $line;
		my $location = $lines[++$ii];

		if($function eq "??" && $location eq "??:0")
		{
			$addr{$sym || $symIdx} = undef;
		}
		else
		{
			my ($f, $l) = split(/:/o, $location, 2);

			$addr{$sym || $symIdx} = { function => $function, file => $f, line => $l };
		}

		$symIdx++;
	}

	return %addr;
}

# process($mapsFile, $workingDir, $pid, @symbols): @resolved_symbols
sub process($$$@)
{
	my ($mapsFile, $wd, $pid, @symbols) = @_;

	# get wordkir
	my $cwd = get_wd($wd, $pid);

	# get library map from maps file
	my (@maps, %libs);
	@maps = read_mapsFile($mapsFile, $pid);
	if(@maps)
	{
		%libs = get_libmap(@maps);
		return wantarray ? (undef, "failed to get library maps") : undef
			if(!keys(%libs));
	}

	# parse input
	my @data;
	foreach my $sym (@symbols)
	{
		my $retry = 0;

		chomp($sym);
		push(@data, $sym);	# as fallback return input symbol

		#return wantarray ? (undef, "failed to parse symbol '$sym'") : undef
		next if($sym eq "[0x0]");
		verbose(1, "SYM_PARSE_FAILED: $sym\n"), next
			if(!($sym =~ /^(.*?)\((.*?)\)?\[(.*?)\]$/o));
		my ($exe, $offset, $addr) = ($1, $2, $3);

		my $exeBN = basename($exe);

		# try to find absolute path via maps file
		$exe = $libs{'-' . $exeBN}
			if(exists($libs{'-' . $exeBN}));

		# not asolute path yet (try it with work-dir)
		my $exeOld = $exe;
		if($cwd && $exe !~ /^\//o)
		{
			my $exe2 = $cwd . '/' . $exe;
			$exe2 = Cwd::abs_path($exe2);

			verbose(1, "EXE_ABSPATH_FAILED: $exe\n"), next
				if(!$exe2 || ! -e $exe2);
			$exe = $exe2;
		}

		#return wantarray ? (undef, "executable '$exeOld' is not a plain file") : undef
		verbose(1, "EXE_NOT_FOUND: $exeOld\n"), next
			if(!$exe || !-f $exe);

		# need real file not symlink
		$exeOld = $exe;
		while(defined($exe) && -l $exe && $retry++ < 10)
		{
			my $dir = dirname($exe);

			$exeOld = $exe;
			$exe = readlink($exe);

			$exe = $dir . '/' . $exe
				if($dir && $exe && $exe !~ /^\//o);
		}

		#return wantarray ? (undef, "failed to read symlink '$exeOld' - $!") : undef
		verbose(1, "EXE_SYMLINK_FOLLOW_FAILED: $exeOld\n"), next
			if(!defined($exe));

		# get absolute path
		$exe = Cwd::abs_path($exe);

		#return wantarray ? (undef, "executable '$exe' is not a plain file") : undef
		verbose(1, "EXE_NOT_FOUND: $exe\n"), next
			if(!-f $exe);

		# library offset via basename
		$exeBN = basename($exe);
		my $libOffset = 0;
		$libOffset = $libs{$exeBN}
			if(exists($libs{$exeBN}));

		# addr offset
		no warnings 'portable';
		my $addrNum = hex($addr);
		my $addrReal = $addrNum - $libOffset;
		my $addrHex = sprintf("0x%02x", $addrReal);

		# success
		pop(@data);
		push(@data, { exe => $exe, offset => $offset, addr_sym => $addr,
				addr_num => $addrNum, addr_real => $addrReal,
				addr_hex => $addrHex,
				sym => $sym });
	}

	# group translate
	my @lines;
	for(my $ii = 0; $ii <= $#data; $ii++)
	{
		my $rec = $data[$ii];

		push(@lines, $rec), next
			if(!ref($rec) || !exists($rec->{exe}));

		# collect all addr from same exe
		my %translate = ( $rec->{addr_hex} => $ii);
		for(my $oo = $ii; $oo <= $#data; $oo++)
		{
			next
				if(!ref($data[$oo])
					|| !exists($data[$oo]->{exe})
					|| $data[$oo]->{exe} ne $rec->{exe});
			$translate{ $data[$oo]->{addr_hex} } = $oo;
		}

		# translate from exe
		my %translated = addr2line($rec->{exe}, keys(%translate));
		push(@lines, $rec), next
			if(!keys(%translated));

		# update data list
		foreach my $addr (keys(%translated))
		{
			my $idx = $translate{ $addr };
			die("[programming error]")
				if(!defined($idx));

			# not translated
			$data[$idx] = $data[$idx]->{sym}, next
				if(!$translated{ $addr });

			my $sym = $data[$idx]->{sym};
			$data[$idx] = $translated{ $addr };
			$data[$idx]->{sym} = $sym;
		}

		# current record translated
		if(defined($data[$ii]))
		{	push(@lines, $data[$ii]);	}
		else	# resolve failed
		{	push(@lines, $rec->{sym});	}
	}

	return @lines;
}

# checkASLR(): bool
sub checkASLR()
{
	my $enabled = 0;

	# LINUX randomization check
	if(-e $LINUX_ALSR_CONFIG
		&& open(my $fd, $LINUX_ALSR_CONFIG))
	{
		my $value = <$fd>;

		close($fd);

		# value is non-zero
		# 0 - no randomization
		# 1 - randomize libs, stack, mmap, VDSO, heap
		# 2 - also brk()
		$enabled = ($value != 0);
	}
	return $enabled;
}

#
# MAIN
#

sub main(@)
{
	my (@argv) = @_;
	my ($exe, $pid, $mapsFile, $workDir, @symbols, $fullName, $man, $help);

	@ARGV = @argv;
	GetOptions(
		"<>"		=> sub { push(@symbols, "$_[0]"); },
		"p|pid=i"	=> \$pid,
		"m|maps-file=s"	=> \$mapsFile,
		"wd|work-dir=s"	=> \$workDir,
		"full-filename"	=> \$fullName,
		"demangle=s"	=> sub { push(@ADDR2LINE, "--demangle=$_[1]"); },
		"v|verbose"	=> \$VERBOSE,
		"h|?|help"	=> \$help,
		"man"		=> \$man,
	) || pod2usage( -verbose => 0,
		-exitval => 1);
	@argv = @ARGV;

	pod2usage( -verbose => 1)
		if($help);
	pod2usage( -verbose => 3)
		if($man);

	# read from STDIN or from file
	if(!@symbols || 
		(my $symFile = ($#symbols == 0 && -e $symbols[0])))
	{
		my $fd = \*STDIN;

		die("$0: failed to open symbols file '$symbols[0]' - $!\n")
			if($symFile && $symbols[0] ne "-" && !open($fd, $symbols[0]));

		# read symbols
		@symbols = _readFile($fd);

		# read maps
		if(!$mapsFile || ($mapsFile && $mapsFile eq "-"))
		{
			$mapsFile = [];
			@$mapsFile = _readFile($fd);
		}

		close($fd);
	}

	# ASLR check
	warn("WARNING: ASLR enabled, but no --pid or --maps-file provided !\n")
		if(checkASLR() && (!$pid && !$mapsFile));

	# process symbols now
	my @data = process($mapsFile, $workDir, $pid, @symbols);

	# error
	die("$0: $data[1]\n")
		if(!defined($data[0]));

	foreach my $rec (@data)
	{
		if(ref($rec))
		{
			my $file = $rec->{file};
			$file = basename($file)
				if(!$fullName);

			printf("%s at %s:%s\n", $rec->{function}, $file, $rec->{line});
		}
		else
		{
			printf("%s (TRANSLATE FAILED)\n", $rec);
		}
	}

	return 0;
}

1;

=pod

=head1 NAME

backtrace2line - convert libc() backtrace() output into file names and line numbers.

=head1 SYNOPSIS

backtrace2line [ OPTIONS ] [ I<SYMBOL-1> ...  I<SYMBOL-N> ]

backtrace2line [ OPTIONS ] I<BACKTRACE-FILE>

=head1 DESCRIPTION

This script converts output of backtrace_symbols() or backtrace_symbols_fd() into file names and line numbers.
It also supports conversion of backtraces produces on machines with an active ASLR (Address space layout randomization),
if pid or process memory map is provided (/proc/<pid>/maps).

NOTE: This script can be also used as perl module.

=head1 ARGUMENTS

=over 4

=item I<BACKTRACE-FILE>

Path to backtrace file. This file must contain symbols (one per line) and I<might> contain also content of 
the maps file, divided by line with single comma (see B<EXAMPLES>). The same applies if reading backtrace
from stdin.

=item I<SYMBOL>

Symbol as generated by backtrace_symbols() function (ie. ./leak-01(main+0x40)[0x7f00ad7a4b00]).

=back

=head1 OPTIONS

=over 4

=item B<-p> I<PID>

=item B<--pid> I<PID>

Pid of a B<still> running process, that generated given backtrace (needed only if ASLR active). Pid
is esed to get process memory map (from /proc/<pid>/maps).

=item B<-m> I<MAPS-FILE>

=item B<--maps-file> I<MAPS-FILE>

Path to a file with process memory map of a backtraced process. This is required if if ALSR is
active. Data from maps file are also used to find full path of the application/libraries from
symbol(s).

=item B<-wd> I<WORK-DIR>

=item B<--work-dir> I<WORK-DIR>

Original work or start dir of backtraced process (needed only if backtrace contains relative paths,
and maps files has not been provided).

=item B<--demangle> I<STYLE>

Passes given --demangle I<STYLE> parameter to B<addr2line> when translating symbols.

=item B<-v>

=item B<--verbose>

Enable verbose mode. This can help you to findout why symbols translation failed.

=item B<-h>

=item B<--help>

Print help.

=item B<--man>

Show man page.

=back

=head1 EXAMPLES

	# pass symbols and maps divided by single comma on line
	$ backtrace2line
	./leak-01(main+0x32)[0x7f00ad7a4af2]
	/lib64/libc.so.6(__libc_start_main+0x11b)[0x7f00acfe31cb]
	./leak-01(+0x9a9)[0x7f00ad7a49a9]
	[0x0]
	.
	7f00ac983000-7f00ac999000 r-xp 00000000 fe:02 911050                     /usr/lib64/gcc/x86_64-pc-linux-gnu/4.8.3/libgcc_s.so.1
	7f00ac999000-7f00acb98000 ---p 00016000 fe:02 911050                     /usr/lib64/gcc/x86_64-pc-linux-gnu/4.8.3/libgcc_s.so.1
	7f00acb98000-7f00acb99000 r--p 00015000 fe:02 911050                     /usr/lib64/gcc/x86_64-pc-linux-gnu/4.8.3/libgcc_s.so.1
	7f00acb99000-7f00acb9a000 rw-p 00016000 fe:02 911050                     /usr/lib64/gcc/x86_64-pc-linux-gnu/4.8.3/libgcc_s.so.1
	7f00acb9a000-7f00acb9d000 r-xp 00000000 fe:02 819448                     /lib64/libdl-2.19.so
	7f00acb9d000-7f00acd9c000 ---p 00003000 fe:02 819448                     /lib64/libdl-2.19.so
	7f00acd9c000-7f00acd9d000 r--p 00002000 fe:02 819448                     /lib64/libdl-2.19.so
	7f00acd9d000-7f00acd9e000 rw-p 00003000 fe:02 819448                     /lib64/libdl-2.19.so
	7f00acd9e000-7f00acdb8000 r-xp 00000000 fe:02 818913                     /lib64/libpthread-2.19.so
	7f00acdb8000-7f00acfb8000 ---p 0001a000 fe:02 818913                     /lib64/libpthread-2.19.so
	7f00acfb8000-7f00acfb9000 r--p 0001a000 fe:02 818913                     /lib64/libpthread-2.19.so
	7f00acfb9000-7f00acfba000 rw-p 0001b000 fe:02 818913                     /lib64/libpthread-2.19.so
	7f00acfba000-7f00acfbe000 rw-p 00000000 00:00 0
	7f00acfbe000-7f00ad170000 r-xp 00000000 fe:02 819451                     /lib64/libc-2.19.so
	7f00ad170000-7f00ad370000 ---p 001b2000 fe:02 819451                     /lib64/libc-2.19.so
	7f00ad370000-7f00ad374000 r--p 001b2000 fe:02 819451                     /lib64/libc-2.19.so
	7f00ad374000-7f00ad376000 rw-p 001b6000 fe:02 819451                     /lib64/libc-2.19.so
	7f00ad376000-7f00ad37a000 rw-p 00000000 00:00 0
	7f00ad37a000-7f00ad37e000 r-xp 00000000 fe:02 1101433                    /usr/local/lib64/liblog-malloc2.so.1.0.0
	7f00ad37e000-7f00ad57d000 ---p 00004000 fe:02 1101433                    /usr/local/lib64/liblog-malloc2.so.1.0.0
	7f00ad57d000-7f00ad57e000 r--p 00003000 fe:02 1101433                    /usr/local/lib64/liblog-malloc2.so.1.0.0
	7f00ad57e000-7f00ad57f000 rw-p 00004000 fe:02 1101433                    /usr/local/lib64/liblog-malloc2.so.1.0.0
	7f00ad57f000-7f00ad5a1000 r-xp 00000000 fe:02 818908                     /lib64/ld-2.19.so
	7f00ad79c000-7f00ad7a1000 rw-p 00000000 00:00 0
	7f00ad7a1000-7f00ad7a2000 r--p 00022000 fe:02 818908                     /lib64/ld-2.19.so
	7f00ad7a2000-7f00ad7a3000 rw-p 00023000 fe:02 818908                     /lib64/ld-2.19.so
	7f00ad7a3000-7f00ad7a4000 rw-p 00000000 00:00 0
	7f00ad7a4000-7f00ad7a5000 r-xp 00000000 fe:02 10097                      /home/sam/devel/log-malloc2/examples/leak-01
	7f00ad9a4000-7f00ad9a5000 r--p 00000000 fe:02 10097                      /home/sam/devel/log-malloc2/examples/leak-01
	7f00ad9a5000-7f00ad9a6000 rw-p 00001000 fe:02 10097                      /home/sam/devel/log-malloc2/examples/leak-01
	7f00af5ac000-7f00af5cd000 rw-p 00000000 00:00 0                          [heap]
	7ffffbe14000-7ffffbe36000 rw-p 00000000 00:00 0                          [stack]
	7ffffbef5000-7ffffbef7000 r-xp 00000000 00:00 0                          [vdso]
	ffffffffff600000-ffffffffff601000 r-xp 00000000 00:00 0                  [vsyscall]
	.
	main at leak-01.c:7
	/lib64/libc.so.6(__libc_start_main+0x11b)[0x7f00acfe31cb] (TRANSLATE FAILED)
	_start at ??:?
	[0x0] (TRANSLATE FAILED)

	$ backtrace2line "./leak-01(main+0x32)[0x7f00ad7a4af2]" "/lib64/libc.so.6(__libc_start_main+0x11b)[0x7f00acfe31cb]" -v
	WARNING: ASLR enabled, but no --pid or --maps-file provided !
	EXE_NOT_FOUND: ./leak-01
	./leak-01(main+0x32)[0x7f00ad7a4af2] (TRANSLATE FAILED)
	/lib64/libc.so.6(__libc_start_main+0x11b)[0x7f00acfe31cb] (TRANSLATE FAILED)

=head1 LICENSE

This script is released under GNU GPLv3 License.
See L<http://www.gnu.org/licenses/gpl.html>.

=head1 AUTHOR

Samuel Behan - L<http://devel.dob.sk/log-malloc2/>, L<https://github.com/samsk/log-malloc2>

=head1 SEE ALSO

L<addr2line>, L<log-malloc>

=cut

# EOF
